packages/ses/src/module-proxy.js

```diff
@@ -0,0 +1,137 @@
+
+import { makeAlias } from './module-load.js';
+
+const { create, freeze } = Object;
+// q, as in quote, for error messages.
+const q = JSON.stringify;
+
+// `deferExports` creates a module's exports proxy, proxied exports, and
+// activator.
+// A `Compartment` can create a module for any module specifier, regardless of
+// whether it is loadable or executable, and use that object as a token that
+// can be fed into another compartment's module map.
+// Only after the specified module has been analyzed is it possible for the
+// module namespace proxy to behave properly, so it throws exceptions until
+// after the compartment has executed the module.
+// The module instance must develop the proxied exports during execution,
+// seal them, and only then activate the module proxy.
+export const deferExports = () => {
+  let active = false;
+  const proxiedExports = create(null);
+  return freeze({
+    activate() {
+      active = true;
+    },
+    proxiedExports,
+    exportsProxy: new Proxy(proxiedExports, {
+      get(_target, name, receiver) {
+        if (!active) {
+          throw new TypeError(
+            `Cannot get property ${q(
+              name,
+            )} of module exports namespace, the module has not yet begun executing`,
+          );
+        }
+        return Reflect.get(proxiedExports, name, receiver);
+      },
+      set(_target, name, _value) {
+        throw new TypeError(
+          `Cannot set property ${q(name)} of module exports namespace`,
+        );
+      },
+      has(_target, name) {
+        if (!active) {
+          throw new TypeError(
+            `Cannot check property ${q(
+              name,
+            )}, the module has not yet begun executing`,
+          );
+        }
+        return Reflect.has(proxiedExports, name);
+      },
+      deleteProperty(_target, name) {
+        throw new TypeError(
+          `Cannot delete property ${q(name)}s of module exports namespace`,
+        );
+      },
+      ownKeys(_target) {
+        if (!active) {
+          throw new TypeError(
+            `Cannot enumerate keys, the module has not yet begun to execute`,
+          );
+        }
+        // return Object.keys(proxiedExports);
+        return Reflect.ownKeys(proxiedExports);
+      },
+      getOwnPropertyDescriptor(_target, name) {
+        if (!active) {
+          throw new TypeError(
+            `Cannot get own property descriptor ${q(
+              name,
+            )}, the module has not yet begun to execute`,
+          );
+        }
+        return Reflect.getOwnPropertyDescriptor(proxiedExports, name);
+      },
+      preventExtensions(_target) {
+        if (!active) {
+          throw new TypeError(
+            `Cannot prevent extensions of module exports namespace, the module has not yet begun to execute`,
+          );
+        }
+        return Reflect.preventExtensions(proxiedExports);
+      },
+      isExtensible() {
+        return Reflect.isExtensible(proxiedExports);
```

Comment:
Interesting that you don't error this one if `!active`. Why break the pattern?

If asked early, does this return `true`? I don't see how to avoid that if it isn't an error. I prefer the error.

Reply:
It would return true before the module finishes executing. This does need to be an error before activate.